designs/autoscaler.md

```diff
@@ -0,0 +1,515 @@
+
+### API Design
+
+```go
+type WorkerNodeGroupConfiguration struct {
+    // Count defines the number of desired worker nodes. Defaults to 1.
+    Count int `json:"count,omitempty"`
+    
+    // AutoScalingConfiguration defines the auto scaling configuration
+    // +optional
+    AutoScalingConfiguration *AutoScalingConfiguration
+    ...
+}
+
+type AutoScalingConfiguration struct {
+    // MinCount defines the minimum number of nodes for the associated resource group.
+    // +optional
+    MinCount int `json:"minCount,omitempty"`
+    
+    // MaxCount defines the maximum number of nodes for the associated resource group.
+    // +optional
+    MaxCount int `json:"maxCount,omitempty"`
+}
+```
+
+### `AutoScalingConfiguration` Validation
+
+If `AutoScalingConfiguration` is not specified, we won’t add any annotation to the underlying MachineDeployment for the worker node group. Instead, we rely on the worker node group count to provision fixed number of nodes.
+
+When `AutoScalingConfiguration` is specified, we validate
+
+* minCount > 0, since Cluster Autoscaler with CAPI does not yet support scaling down to 0
+* minCount <= count <= maxCount
+
+### CAPI  Machine Deployment
+
+Internally, the `minCount` and `maxCount` defined in the EKS-A cluster spec is converted to the corresponding worker node group MachineDeployment’s annotations `cluster.x-k8s.io/cluster-api-autoscaler-node-group-max-size` and `cluster.x-k8s.io/cluster-api-autoscaler-node-group-min-size`, as EKS-A builds the CAPI templates. The Cluster Autoscaler will monitor any of those MachineDeployment containing both of these annotations.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDeployment
+metadata:
+  annotations:
+    cluster.x-k8s.io/cluster-api-autoscaler-node-group-max-size: {{.maxCount}}
+    cluster.x-k8s.io/cluster-api-autoscaler-node-group-min-size: {{.minCount}}
+    cluster.x-k8s.io/cluster-name: {{.clusterName}}
+  name: {{.workerNodeGroupName}}
+  namespace: eksa-system
+spec:
+  clusterName: {{.clusterName}}
+  replicas: {{.count}}
+  ...
+```
+
+Note: the cluster will not have any autoscaling functionality until the Cluster Autoscaler controller is properly configured and installed in the cluster. Setting up the MachineDeployment annotations only guarantees that Cluster Autoscaler can recognize the worker node group and its autoscaling configuration. We will add a warning message in CLI logs when the autoscaling is configured in EKS-A cluster spec but autoscaler is not installed in the cluster.
+
+### Cluster Autoscaler Controller
+
+Cluster Autoscaler is a tool that automatically adjusts the size of the Kubernetes cluster. Specifically, it can:
+
+* increase the size of the cluster when there are pending pods due to insufficient resources.
+* decrease the size of the cluster when some nodes have low utilization for a significant amount of time.
```

Comment:
how are "insufficient resources" and "low utilization" defined? Is this something a user would be able to control? Things like time durations, specific metrics to monitor (avg vs max utilization), etc.

Reply:
details are covered here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-down-work
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-are-expanders

some metric you can configure in CA https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-are-the-parameters-to-ca